### ISSUE TITLE
Assigning IPV6 Localhost address ::1 to RemoteAddresses throws exception

### ISSUE THREAD

**Issue by TimesliceTechnologies** [falahati/WindowsFirewallHelper#8]
When one of the addresses in the array of IAddress that is assigned to Rule.RemoteAddresses is ::1 an exception is thrown. ::1 is IPV6 for LocalHost.
IPV4 127.0.0.1 works fine.

**Example assignment.  NewIPs is an array of IAddress:**
objRule.RemoteAddresses = NewIPs.ToArray();

**Here's the list of IP addresses in NewIPs.  Note the last one is ::1**
IP addresses:173.35.205.201,99.248.70.97,66.96.146.81,::1

**Here's the exception and stacktrace:**

Exception from HRESULT: 0xD000000D   at NetFwTypeLib.INetFwRule.set_RemoteAddresses(String remoteAddrs)
   at WindowsFirewallHelper.FirewallAPIv2.Rules.StandardRule.set_RemoteAddresses(IAddress[] value)
   at FWManagerService.FirewallJobProcessor.ApplyAuthorizedRemoteAddresses(List`1 lstAuthIPs) in D:\VBapps\FWmanager\FWManagerService\Code\FirewallJobProcessor.cs:line 291

**Comment by falahati** [comment]
This is a Windows Firewall limitation. You can not specify any loopback, broadcast or multicast IPv6 address. That being said, even using `127.0.0.1` as the remote or local address is not going to work as Windows Firewall can't intercept traffics that not going to get routed.

What should I do in the library to make this limitation clear is another issue though.

**Comment by TimesliceTechnologies** [comment]
Maybe something minimal like throwing a more descriptive exception if any address in the list contains a loopback or other non-allowed address?

Maybe throw an exception that uses similar message as the Window Firewall GUI:
"An unspecified, multicast, broadcast or loopback IPv6 address was specified"
 
I will filter these out in my own front end to avoid the error for now...

**Comment by falahati** [comment]
If your project ended up being open source, it would be great to add a link to its source code inside the readme file for others to see a working example.

**Comment by TimesliceTechnologies** [comment]
I will consider that.  It is currently an in house windows service project used to lockdown RDP access to servers with a related web app that allows adjusting port rules remotely via a WebApi in the service.  The service monitors DNS changes and some other internal and external Host lists and leverages WIndowsFirewallHelper to sync the firewall rule to the allowed addresses.

I have considered releasing the final RDP lockdown product as an inexpensive commercial product to offset my development time/costs.  Open sourcing it hadn't really occurred to me.  I'd be most interested in your experience with open sourcing and it's rewards (financial or otherwise).

WIndowsFirewallHelper is exceptionally useful and well written and I am most thankful for your quick fixes and assistance.

**Comment by falahati** [comment]
> I'd be most interested in your experience with open sourcing and it's rewards (financial or otherwise)

Not much of a reward to be honest. I go for open source in meta projects that I might never get back to or use anywhere else. It a way for me to let the code live. Especially libraries that others may find useful. It takes some time as it needs to be documented and all to be released as an open source project, but at least it gives me the feeling that the time I spend on the code, was not limited to the project I was working on and others may use the code for other similar porpuses. This is why I rarely open source any end-user code and ready to use applications. The only "reward" I got was how other developers see my libraries and know what I can do and what I did in the past. So we can work better together as a team knowing each other better.

> I have considered releasing the final RDP lockdown product as an inexpensive commercial product to offset my development time/costs.

Even if you do that, having a link to the product page in the readme would be great. It is still an example of a commercial product using the library.

This is an MIT project, so you are free to use it anywhere you want and even make a private fork and all. You are not bound by the license to release your changes to the library, but I ask you that if you changed something inside of the library that makes it more useful, or less buggy, create a push request. Thanks

**Comment by TimesliceTechnologies** [comment]
Thanks for the insight.  Agreed on your motivation for open sourcing some of your work.
I will keep you posted regarding my final product.  

I'm about to deploy on several of our public facing servers to see how it behaves.  I look forward to locking down RDP access against the thousands of brute force attacks we have been seeing lately.  Easy to do if your RDP client PCs are at a fixed IP with a simple firewall rule.  Not so easy when the PCs are on dynamic IPs.  My new service provides a way to remotely and automatically control the firewall remotely and auto-sync an entire collection of servers with my own list of allowed addresses.

I have restricted my use of WindowsFirewallHelper to your official release without any changes - I prefer not to fork/modify so that I can easily stay with a standardized version.  Your quick attention to those few IPv6 problems resolved any issues I encountered.

Regarding "rewards" I completely understand how some positive reviews from peers is often sufficient and suitably gratifying.  You have put together a nicely structured and well written component, and your support has been rapid and efficient.

I will provide a link if we turn this into commercial product.  I think there are many folks that have the same issue locking down / managing RDP that I encountered and would find this most useful.
